Refactor Sample Data Home section which also fixes flaky sample data functional test

nreese opened the pull request:
fixes https://github.com/elastic/kibana/issues/20738

To consistently reproduce the flaky test in master, wrap the call to [onRequestComplete](https://github.com/elastic/kibana/blob/master/src/core_plugins/kibana/public/home/components/sample_data_set_card.js#L67) in a timeout like this `setTimeout(onRequestComplete, 1000);`. 

The timing problem causing the flaky test occurs because `SampleDataSetCard` `status` prop would occasionally not get updated before the expect clause gets executed.

This PR fixes the timing problem by 
1) refactoring `SampleDataSetCard` to track status in state and avoid the whole mess of having to call `onRequestComplete` and then waiting for `props` to force a render.
2) refactoring the functional tests to only exit `addSampleDataSet` and `removeSampleDataSet` when the loading indicator is gone.

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/8803/)

Stacey-Gammon commented on src/core_plugins/kibana/public/home/components/sample_data_set_card.js:
```diff
@@ -58,36 +67,53 @@ export class SampleDataSetCard extends React.Component {
+    } = this.props;
+
+    this.setState({
+      isProcessingRequest: true,
```
This is just another status.  Can we merge the props into one?  Status = PROCESSING, INSTALLED, UNINSTALLED.  and I guess maybe UNKOWN?

How do I simulate UNKNOWN?  I can't seem to trigger it.

I think even if you handle the isMounted checks below you can get into a weird state by hitting 'Add' or 'Remove' the quickly switching tabs and going back.  isProcessingREquest is state on this button so that gets lost but the request is still processing.  Then I think you will end up with indicating the wrong status.

<img width="1248" alt="screen shot 2018-08-06 at 2 01 50 pm" src="https://user-images.githubusercontent.com/16563603/43733032-51ee06d2-9981-11e8-9554-d583bea118bf.png">

nreese replied:
`UNKNOWN ` surfaces if there are any problems accessing the data, like permission problems

nreese commented:
@Stacey-Gammon I moved all of the sample data logic from the `TutorialDirectory` component and into a new parent component, `SampleDataSetCards`. I removed state from `SampleDataSetCard` so that all state management is know in the new parent component.

I could not add `processing` as another state in `status`. Processing can occur in both the `installed` and `not_installed` status so it needs to be its own value.

Stacey-Gammon commented on src/core_plugins/kibana/public/home/components/sample_data_set_cards.js:
```diff
@@ -0,0 +1,167 @@
+    this.setState({
+      sampleDataSets: sampleDataSets
+        .map((sampleDataSet) => {
+          sampleDataSet.isProcessing = false;
```
Looks like this is modifying the returned object.  When we have typescript, the structure of `sampleDataSet` will be easy to see, and at that point, I don't think it would make sense to put an `isProcessing` flag on it just for the purpose of this class.  What about a mapping that keeps track of `isProcessing` for each sample data set by id?

Stacey-Gammon commented on src/core_plugins/kibana/public/home/components/sample_data_set_card.js:
```diff
@@ -28,70 +28,40 @@ import {
-  installSampleDataSet,
-  uninstallSampleDataSet
-} from '../sample_data_sets';
+const INSTALLED_STATUS = 'installed';
```
I guess these are returned from the objects in `listSampleDataSets`?  Our folder structure of core plugins makes it difficult to push common consts in an intuitive place (the server side code of this is far away).   If we typescripted this and the files in `../sample_data_sets` we could at least show that the object being returned has a `status` and is one of `installed` or `not_installed`.  Suppose that's fine for this PR but I think it'd be really helpful to understand and follow this code if it was in typescript.

nreese commented:
The test failed because of another async problem. The toast notifications where getting generated within the functions that handle calling the REST API. This could have a timing problem if the toast gets displayed and then the UI does not update because `loadSampleDataSets` takes longer than the toast notification is displayed. 

The latest commit moves the toast notification generation to the component. Now they are only displayed after `loadSampleDataSets` has finished so there can be no more timing errors.

Also, the functions that call the sample data REST APIs have been updated to use `kfetch` which removes a bunch of redundant error checking code

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/450/)

nreese commented:
Not sure why the selenium test failed, looks like elasticsearch could not start

```
15:49:54    │ERROR  Note: /var/lib/jenkins/workspace/elastic+kibana+pull-request+multijob-selenium/elasticsearch/libs/x-content/src/main/java/org/elasticsearch/common/xcontent/json/JsonXContentGenerator.java uses or overrides a deprecated API.
15:49:54    │ERROR  Note: Recompile with -Xlint:deprecation for details.
15:49:54    │ debg  > Task :libs:x-content:jar
15:50:24    │ debg  > Task :server:compileJava
15:50:25    │ERROR  Note: Some input files use or override a deprecated API.
15:50:25    │ERROR  Note: Recompile with -Xlint:deprecation for details.
15:50:25    │ERROR  Note: Some input files use unchecked or unsafe operations.
15:50:25    │ERROR  Note: Recompile with -Xlint:unchecked for details.
```

jenkins, test this

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/455/)

nreese commented:
flaky dashboard filtering test

```
fail: "dashboard app using current data dashboard filtering adding a filter that excludes all data "before all" hook"
17:23:28    │ proc  [ftr]          │        retry.try timeout: Error: 16 elements completed rendering, waiting on a total of 17
```

jenkins, test this

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/466/)

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/508/)

nreese commented:
jenkins, test this

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/539/)

nreese commented:
jenkins, test this

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/548/)

nreese commented:
dashboard app test failure

```
fail: "dashboard app using current data create and add embeddables "before all" hook"
16:07:14    │ proc  [ftr]        │        retry.try timeout: Error: retry.try timeout: Error: retry.try timeout: [POST http://localhost:9515/session/645431cafa2c24a930be622dd01fb2d4/element / {"using":"css selector","value":"a[href=\"#/dashboards\"]"}] no such element: Unable to locate element: {"method":"css selector","selector":"a[href="#/dashboards"]"}
```

jenkins, test this

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/552/)

nreese commented:
jenkins, test this

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/561/)

nreese commented:
jenkins, test this

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/576/)

nreese commented:
jenkins, test this

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/622/)

nreese commented:
looks like an unrelated error

```
 ERROR  [POST http://localhost:9515/session / {"desiredCapabilities":{},"requiredCapabilities":{}}] unknown error: Chrome failed to start: exited abnormally
14:16:02    │ proc  [ftr]          (unknown error: DevToolsActivePort file doesn't exist)
14:16:02    │ proc  [ftr]          (The process started from chrome location /usr/bin/google-chrome is no longer running, so ChromeDriver is assuming that Chrome has crashed.)
14:16:02    │ proc  [ftr]          (Driver info: chromedriver=2.41.578700 (2f1ed5f9343c13f73144538f15c00b370eda6706),platform=Linux 4.9.0-7-amd64 x86_64)
14:16:02    │ proc  [ftr]            at Server._post (/var/lib/jenkins/workspace/elastic+kibana+pull-request+multijob-selenium/kibana/test/functional/services/remote/verbose_remote_logging.js:34:21)
14:16:02    │ proc  [ftr]            at Server.createSession (/var/lib/jenkins/workspace/elastic+kibana+pull-request+multijob-selenium/kibana/node_modules/leadfoot/Server.js:362:15)
14:16:02    │ proc  [ftr]            at attemptToCreateCommand (/var/lib/jenkins/workspace/elastic+kibana+pull-request+multijob-selenium/kibana/test/functional/services/remote/leadfoot_command.js:38:32)
14:16:02    │ proc  [ftr]            at Promise.race (/var/lib/jenkins/workspace/elastic+kibana+pull-request+multijob-selenium/kibana/test/functional/services/remote/leadfoot_command.js:86:11)
14:16:02    │ proc  [ftr]            at initLeadfootCommand (/var/lib/jenkins/workspace/elastic+kibana+pull-request+multijob-selenium/kibana/test/functional/services/remote/leadfoot_command.js:63:5)
14:16:02    │ proc  [ftr]            at RemoteProvider (/var/lib/jenkins/workspace/elastic+kibana+pull-request+multijob-selenium/kibana/test/functional/services/remote/remote.js:40:29)
14:16:02    │ proc  [ftr]            at <anonymous>
14:16:02    │ proc  [ftr] ERROR  Error: Failure initializing 1 service(s)
14:16:02    │ proc  [ftr]            at ProviderCollection.loadAll (/var/lib/jenkins/workspace/elastic+kibana+pull-request+multijob-selenium/kibana/src/functional_test_runner/lib/providers/provider_collection.js:67:13)
14:16:02    │ proc  [ftr]            at <anonymous>
14:16:02    │ proc  [ftr]            at process._tickCallback (internal/process/next_tick.js:188:7)
14:16:02    │ warn  1 processes left running, stop them with procs.stop(name): [ 'kibana' ]
14:16:02    │ info  [kibana] exited with null after 2 minutes
```

jenkins, test this

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/632/)

nreese commented:
jenkins, test this

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/662/)

nreese commented:
@Stacey-Gammon This is ready for a final review

Stacey-Gammon commented on src/core_plugins/kibana/public/home/components/sample_data_set_cards.js:
```diff
@@ -0,0 +1,191 @@
+      return;
+    }
+
+    await this.loadSampleDataSets();
```
This is hard to verify with only one actual sample data set available, but I think an error could appear if the client hit add consecutively on multiple data sets.  It appears that the first response from `installSampleDataSet` will call `loadSampleDataSets` which clears out all processingStatus states.

nreese replied:
should the local processingStatus be preserved?

Stacey-Gammon replied:
I just suspect a bug will arise when we add more sample data sets, where if you pressed "Add", "Add", "Add", all three states go to "true", so then you'd see "Adding", "Adding", "Adding".  When the first one returns, it'll set all three to false.  Reloading the data might show that the one was successfully added, but the others, may not have finished processing, so I think now you'd see "Remove", "Add", "Add".  Even though the other two are still in the process of being added.

nreese commented:
@Stacey-Gammon I updated the logic to only load sample data set status on initial component load. After that, status is tracked in state to avoid the problem you identified.

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/879/)

Stacey-Gammon commented on src/core_plugins/kibana/public/home/components/sample_data_set_cards.js:
```diff
@@ -0,0 +1,211 @@
+    } catch (fetchError) {
+      if (this._isMounted) {
+        this.setState((prevState) => ({
+          processingStatus: { ...prevState.processingStatus, [id]: true }
```
should this be `[id]: false`?

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/928/)
